ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java

```diff
@@ -0,0 +1,182 @@
+
+    Request request = new Request.Builder()
+        .tag(id)
+        .url(url)
+        .build();
+
+    WebSocketCall.create(client, request).enqueue(new WebSocketListener() {
+      @Override
+      public void onOpen(final WebSocket webSocket, final Response response) {
+        mWebSocketConnections.put(id, webSocket);
+
+        WritableMap params = Arguments.createMap();
+        params.putInt("id", id);
+        sendEvent("websocketOpen", params);
+      }
+
+      @Override
+      public void onMessage(
+          final BufferedSource bufferedSource,
+          final WebSocket.PayloadType payloadType
+      ) {
+        String message;
+
+        WritableMap params = Arguments.createMap();
+        params.putInt("id", id);
+
+        try {
+          message = bufferedSource.readUtf8();
+        } catch (IOException e) {
+          params.putString("message", e.getMessage());
+          sendEvent("websocketErrored", params);
+
+          return;
+        }
+
+        params.putString("data", message);
+
+        try {
+          bufferedSource.close();
+        } catch (IOException e) {
+          FLog.e(
+            ReactConstants.TAG,
+            "Could not close BufferedSource for WebSocket message " + id,
+            e);
+        }
+
+        sendEvent("websocketMessage", params);
+      }
+
+      @Override
+      public void onClose(final int code, final String reason) {
+        WritableMap params = Arguments.createMap();
+        params.putInt("id", id);
+        params.putInt("code", code);
+        params.putString("reason", reason);
+        sendEvent("websocketClosed", params);
+      }
+
+      @Override
+      public void onFailure(final IOException e, final Response response) {
+        WritableMap params = Arguments.createMap();
+        params.putInt("id", id);
+        params.putString("message", e.getMessage());
+        sendEvent("websocketFailed", params);
+      }
+
+      @Override
+      public void onPong(final Buffer buffer) {
+      }
+    });
+
+    // Trigger shutdown of the dispatcher's executor so this process can exit cleanly.
+    client.getDispatcher().getExecutorService().shutdown();
+  }
+
+  @ReactMethod
+  public void close(final int code, final String reason, final int id) {
+    WebSocket client = mWebSocketConnections.get(id);
+
+    if (client == null) {
+      FLog.w(
+        ReactConstants.TAG,
+        "Could not find WebSocket connection for " + id);
+      return;
+    }
+
+    try {
+      client.close(code, reason);
+      mWebSocketConnections.remove(id);
+    } catch (Exception e) {
+      FLog.e(
+        ReactConstants.TAG,
+        "Could not close WebSocket connection for " + id,
+        e);
+    }
+  }
+
+  @ReactMethod
+  public void send(final String message, final int id) {
+    WebSocket client = mWebSocketConnections.get(id);
+
+    try {
+      client.sendMessage(
+        WebSocket.PayloadType.TEXT,
+        new Buffer().writeUtf8(message)
+      );
```

Comment:
nit: the convention is to place the closing bracket at the end of the previous line